File: leetcode/665. Non-decreasing Array.cpp
class Solution {
public:
    bool checkPossibility(vector<int>& nums) {
        int idx = -1;
        for(size_t i = 0; i < nums.size() - 1; ++i){
            if(nums[i] > nums[i + 1]){
                if(idx != -1){
                    return false;
                }
                
                idx = i;
            }
        }
        
        if(idx == -1 || idx == 0 || idx == nums.size() - 2){
            return true;
        }
        
        if(nums[idx + 1] >= nums[idx - 1] || nums[idx + 2] >= nums[idx]){
            return true;
        }
        
        return false;
    }
};

File: leetcode/665. Non-decreasing Array.py
class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        for i in range(len(nums) - 1):
            if nums[i] > nums[i + 1]:
                tmp = nums[i]
                res = True
                if i == 0 or nums[i + 1] >= nums[i - 1]: 
                    nums[i] = nums[i + 1]
                    for j in range(i, len(nums) - 1):
                        if nums[j] > nums[j + 1]:
                            res = False
                            break
                    if res:
                        return True
                    
                nums[i] = nums[i + 1] = tmp
                for j in range(i, len(nums) - 1):
                    if nums[j] > nums[j + 1]:
                        return False
                return True
        return True
        
        
  
  
  class Solution:
    def checkPossibility(self, nums: List[int]) -> bool:
        idx = None
        for i in range(len(nums) - 1):
            if nums[i] > nums[i + 1]:
                if idx != None:
                    return False
                idx = i
        
        if idx == None or idx == 0 or idx == len(nums) - 2:
            return True
        if nums[idx + 1] >= nums[idx - 1]:
            return True
        if nums[idx + 2] >= nums[idx]:
            return True
        return False
  
